packages/beacon-node/src/chain/chain.ts

```diff
@@ -1271,9 +1261,54 @@ export class BeaconChain implements IBeaconChain {
   }
 
   async updateBeaconProposerData(epoch: Epoch, proposers: ProposerPreparationData[]): Promise<void> {
+    const previousValidatorCount = this.beaconProposerCache.getValidatorIndices().length;
+
     for (const proposer of proposers) {
       this.beaconProposerCache.add(epoch, proposer);
     }
+
+    const newValidatorCount = this.beaconProposerCache.getValidatorIndices().length;
+
+    // Only update validator custody if we discovered new validators
+    if (newValidatorCount > previousValidatorCount) {
+      const finalizedCheckpoint = this.forkChoice.getFinalizedCheckpoint();
+      await this.updateValidatorsCustodyRequirement(finalizedCheckpoint);
+    }
+  }
+
+  private async updateValidatorsCustodyRequirement(finalizedCheckpoint: CheckpointWithHex): Promise<void> {
+    if (this.opts.supernode) {
+      // Disable dynamic custody updates for supernodes since they must maintain custody
+      // of all custody groups regardless of validator effective balances
+      return;
+    }
+
+    // Update custody requirement based on finalized state
+    const stateOrBytes = (await this.getStateOrBytesByCheckpoint(finalizedCheckpoint))?.state;
+
+    if (!stateOrBytes) {
+      throw Error(
+        `No finalized state for epoch ${finalizedCheckpoint.epoch} and root ${finalizedCheckpoint.rootHex} to update target custody group count`
+      );
+    }
+
+    const stateValidators =
+      stateOrBytes instanceof Uint8Array
+        ? getValidatorsFromStateBytes(this.config, stateOrBytes)
```

Comment:
most of the time to deserialize a BeaconState is for validators list so we should not load all of them just to load some effective balances
we should implement a function to extract effective balances given a list of validator index

Reply:
maybe better to use `Map<ValidatorIndex, EffectiveBalance` to avoid a big array